File: lib/features/compatibility/domain/entities/compatibility_reveal.dart
/// A single verified Compatibility Reveal point (server or local engine).
enum CompatibilityRevealKind {
  personality,
  questions,
  music,
  relationship,
  preference,
  lifestyle,
}

class CompatibilityRevealPoint {
  const CompatibilityRevealPoint({
    required this.kind,
    required this.messageKey,
    this.messageArgs = const [],
    this.score,
  });

  final CompatibilityRevealKind kind;
  final String messageKey;
  final List<String> messageArgs;
  final int? score;

  factory CompatibilityRevealPoint.fromMap(Map<String, dynamic> raw) {
    return CompatibilityRevealPoint(
      kind: _kindFrom(raw['kind'] as String?),
      messageKey: (raw['messageKey'] as String?) ?? '',
      messageArgs: _stringList(raw['messageArgs']),
      score: raw['score'] is int ? raw['score'] as int : null,
    );
  }

  static CompatibilityRevealKind _kindFrom(String? value) {
    return switch (value) {
      'personality' => CompatibilityRevealKind.personality,
      'questions' => CompatibilityRevealKind.questions,
      'music' => CompatibilityRevealKind.music,
      'relationship' => CompatibilityRevealKind.relationship,
      'preference' => CompatibilityRevealKind.preference,
      'lifestyle' => CompatibilityRevealKind.lifestyle,
      _ => CompatibilityRevealKind.preference,
    };
  }

  static List<String> _stringList(Object? raw) {
    if (raw is! List) {
      return const [];
    }
    return raw.whereType<String>().where((e) => e.isNotEmpty).toList();
  }
}

class CompatibilityRevealBreakdown {
  const CompatibilityRevealBreakdown({
    required this.relationshipScore,
    required this.interestScore,
    required this.lifestyleScore,
    this.questionScore,
    this.musicScore,
    this.communicationScore,
  });

  final int relationshipScore;
  final int interestScore;
  final int lifestyleScore;
  final int? questionScore;
  final int? musicScore;
  final int? communicationScore;

  factory CompatibilityRevealBreakdown.fromMap(Map<String, dynamic> raw) {
    int? optInt(Object? value) => value is int ? value : null;
    return CompatibilityRevealBreakdown(
      relationshipScore: raw['relationshipScore'] is int
          ? raw['relationshipScore'] as int
          : 0,
      interestScore: raw['interestScore'] is int
          ? raw['interestScore'] as int
          : 0,
      lifestyleScore: raw['lifestyleScore'] is int
          ? raw['lifestyleScore'] as int
          : 0,
      questionScore: optInt(raw['questionScore']),
      musicScore: optInt(raw['musicScore']),
      communicationScore: optInt(raw['communicationScore']),
    );
  }
}

/// Match Compatibility Reveal payload from `getMatchCompatibilityReveal`.
class CompatibilityReveal {
  const CompatibilityReveal({
    required this.available,
    required this.overallScore,
    this.isPremium = false,
    this.premiumRequired = false,
    this.points = const [],
    this.breakdown,
    this.reason,
  });

  static const unavailable = CompatibilityReveal(
    available: false,
    overallScore: 0,
  );

  final bool available;
  final int overallScore;
  final bool isPremium;
  final bool premiumRequired;
  final List<CompatibilityRevealPoint> points;
  final CompatibilityRevealBreakdown? breakdown;
  final String? reason;

  bool get showPremiumUpsell => available && premiumRequired && !isPremium;

  factory CompatibilityReveal.fromMap(Map<String, dynamic> raw) {
    final pointsRaw = raw['points'];
    final points = <CompatibilityRevealPoint>[];
    if (pointsRaw is List) {
      for (final item in pointsRaw) {
        if (item is Map) {
          points.add(
            CompatibilityRevealPoint.fromMap(Map<String, dynamic>.from(item)),
          );
        }
      }
    }
    CompatibilityRevealBreakdown? breakdown;
    final breakdownRaw = raw['breakdown'];
    if (breakdownRaw is Map) {
      breakdown = CompatibilityRevealBreakdown.fromMap(
        Map<String, dynamic>.from(breakdownRaw),
      );
    }
    return CompatibilityReveal(
      available: raw['available'] == true,
      overallScore: raw['overallScore'] is int
          ? raw['overallScore'] as int
          : 0,
      isPremium: raw['isPremium'] == true,
      premiumRequired: raw['premiumRequired'] == true,
      points: points,
      breakdown: breakdown,
      reason: raw['reason'] as String?,
    );
  }
}

File: functions/src/compatibilityReveal.ts
import {getApps, initializeApp} from "firebase-admin/app";
import {getFirestore, type DocumentData} from "firebase-admin/firestore";
import {HttpsError, onCall, type CallableRequest} from "firebase-functions/v2/https";
import {calculateCompatibility} from "./compatibility/compatibilityEngine.js";
import {
  buildCompatibilityReveal,
  type CompatibilityRevealResult,
} from "./compatibilityRevealEngine.js";
import {resolveMatchParticipant} from "./compatibilityRevealGate.js";
import {canonicalMatchId} from "./ids.js";
import {isUserPremium} from "./premium.js";
import {relationshipScoreForPair} from "./relationshipMatch.js";
import {musicScoreForPair} from "./spotifyMusic.js";

if (getApps().length === 0) {
  initializeApp();
}

const db = getFirestore();
const enforceAppCheck = process.env.FUNCTIONS_EMULATOR !== "true";
const callableOptions = {
  region: "europe-west1" as const,
  invoker: "public" as const,
  enforceAppCheck,
};

function requireUid(request: CallableRequest): string {
  const uid = request.auth?.uid;
  if (!uid) {
    throw new HttpsError("unauthenticated", "Sign in required.");
  }
  return uid;
}

function requireString(value: unknown, field: string): string {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new HttpsError("invalid-argument", field);
  }
  return value.trim();
}

function goalLabel(raw: unknown): string | null {
  const value = typeof raw === "string" ? raw.trim() : "";
  if (!value) return null;
  const normalized = value.replace(/_/g, "").toLowerCase();
  if (normalized === "longterm") return "longTerm";
  if (normalized === "casual") return "casual";
  if (normalized === "figuringout" || normalized === "figuring_out") {
    return "figuringOut";
  }
  return value;
}

function sameGoal(a: DocumentData, b: DocumentData): boolean {
  const ga = goalLabel(a.relationshipGoal);
  const gb = goalLabel(b.relationshipGoal);
  return ga != null && gb != null && ga === gb;
}

async function requireActiveMatchParticipant(
  matchId: string,
  uid: string,
): Promise<{peerUid: string}> {
  const snap = await db.doc(`matches/${matchId}`).get();
  const data = snap.data() ?? {};
  const gated = resolveMatchParticipant({
    matchId,
    uid,
    matchExists: snap.exists,
    isActive: data.isActive as boolean | undefined,
    userIds: data.userIds,
    canonicalMatchId,
  });
  if (!gated.ok) {
    if (gated.code === "not-found") {
      throw new HttpsError("not-found", "match-not-found");
    }
    if (gated.code === "match-inactive") {
      throw new HttpsError("failed-precondition", "match-inactive");
    }
    if (gated.code === "not-a-participant" || gated.code === "match-id-mismatch") {
      throw new HttpsError("permission-denied", gated.code);
    }
    throw new HttpsError("failed-precondition", gated.code);
  }
  return {peerUid: gated.peerUid};
}

/**
 * Post-match Compatibility Reveal.
 *
 * Free: overall score + up to 2 basic verified reasons (no answer text).
 * Premium: up to 3 reasons + category breakdown.
 * Never fabricates music/question/personality overlaps without real data.
 */
export const getMatchCompatibilityReveal = onCall(
  callableOptions,
  async (request): Promise<CompatibilityRevealResult> => {
    const uid = requireUid(request);
    const matchId = requireString(request.data?.matchId, "matchId");
    const {peerUid} = await requireActiveMatchParticipant(matchId, uid);
    const premium = await isUserPremium(uid);

    const [viewerSnap, peerSnap, relationship, music] = await Promise.all([
      db.doc(`profiles/${uid}`).get(),
      db.doc(`profiles/${peerUid}`).get(),
      relationshipScoreForPair(uid, peerUid),
      musicScoreForPair(uid, peerUid),
    ]);

    if (!viewerSnap.exists || !peerSnap.exists) {
      return {
        available: false,
        overallScore: 0,
        isPremium: premium,
        premiumRequired: !premium,
        points: [],
        breakdown: null,
        reason: "profile_missing",
      };
    }

    const viewer = viewerSnap.data() ?? {};
    const peer = peerSnap.data() ?? {};
    const musicScore =
      music && typeof music.score === "number" && music.score > 0
        ? music.score
        : null;

    const compat = calculateCompatibility({
      viewerProfile: viewer,
      candidateProfile: peer,
      relationship: relationship
        ? {
            score: relationship.score,
            alignedCount: relationship.alignedCount,
            sharedQuestionCount: relationship.sharedQuestionCount,
            topTopics: relationship.topTopics ?? [],
          }
        : null,
      musicScore,
    });

    const label = goalLabel(viewer.relationshipGoal);

    return buildCompatibilityReveal({
      overallScore: compat.overallScore,
      isPremium: premium,
      sharedInterests: compat.sharedInterests,
      sameRelationshipGoal: sameGoal(viewer, peer),
      relationshipGoalLabel: sameGoal(viewer, peer) ? label : null,
      questionAlignedCount: compat.questionAlignedCount,
      questionSharedCount: compat.questionSharedCount,
      questionScore: compat.questionScore,
      questionTopTopics: relationship?.topTopics ?? [],
      musicScore,
      lifestyleScore: compat.lifestyleScore,
      communicationScore: compat.communicationScore,
      breakdown: {
        relationshipScore: compat.relationshipScore,
        interestScore: compat.interestScore,
        lifestyleScore: compat.lifestyleScore,
        questionScore: compat.questionScore,
        musicScore: compat.musicScore,
        communicationScore: compat.communicationScore,
      },
    });
  },
);
